xrootd/client/session.go

```diff
@@ -0,0 +1,277 @@
+	"go-hep.org/x/hep/xrootd/xrdproto"
+	"go-hep.org/x/hep/xrootd/xrdproto/signing"
+	"go-hep.org/x/hep/xrootd/xrdproto/sigver"
+)
+
+// session is a connection to the specific XRootD server
+// which allows to send requests and receive responses.
+// Concurrent requests are supported.
+// Zero value is invalid, Session should be instantiated using NewSession.
+//
+// The session is used by the Client to send requests to the particular server
+// specified by the name and port. If the current server cannot
+// handle a request, it responds with the redirect to the new server.
+// After that, Client obtains a session associated with that server and
+// re-issues the request. Stream ID may be different during these 2 requests
+// because it is used to identify requests among one particular server
+// and is not shared between servers in any way.
+type session struct {
+	cancel           context.CancelFunc
+	conn             net.Conn
+	mux              *mux.Mux
+	protocolVersion  int32
+	signRequirements signing.Requirements
+	seqID            int64
+	mu               sync.RWMutex
+	requests         map[xrdproto.StreamID][]byte
+	client           *Client
+}
+
+func newSession(ctx context.Context, address string, username string, client *Client) (*session, error) {
+	ctx, cancel := context.WithCancel(ctx)
+
+	var d net.Dialer
+	conn, err := d.DialContext(ctx, "tcp", address)
+	if err != nil {
+		cancel()
+		return nil, err
+	}
+
+	sess := &session{cancel: cancel, conn: conn, mux: mux.New(), requests: make(map[xrdproto.StreamID][]byte), client: client}
+
+	go sess.consume(ctx)
+
+	if err := sess.handshake(ctx); err != nil {
+		sess.Close()
+		return nil, err
+	}
+
+	securityInfo, err := sess.Login(ctx, username, "")
+	if err != nil {
+		sess.Close()
+		return nil, err
+	}
+
+	if len(securityInfo.SecurityInformation) > 0 {
+		err = sess.auth(ctx, securityInfo.SecurityInformation)
+		if err != nil {
+			sess.Close()
+			return nil, err
+		}
+	}
+
+	protocolInfo, err := sess.Protocol(ctx)
+	if err != nil {
+		sess.Close()
+		return nil, err
+	}
+
+	sess.signRequirements = signing.New(protocolInfo.SecurityLevel, protocolInfo.SecurityOverrides)
+
+	return sess, nil
+}
+
+// Close closes the connection. Any blocked operation will be unblocked and return error.
+func (sess *session) Close() error {
+	sess.cancel()
+
+	sess.mux.Close()
+	return sess.conn.Close()
+}
+
+func (sess *session) consume(ctx context.Context) {
+	var header xrdproto.ResponseHeader
+	var headerBytes = make([]byte, xrdproto.ResponseHeaderLength)
+
+	for {
+		select {
+		case <-ctx.Done():
+			// TODO: Should wait for active requests to be completed?
+			return
+		default:
+			if _, err := io.ReadFull(sess.conn, headerBytes); err != nil {
+				if ctx.Err() != nil {
+					// something happened to the context.
+					// ignore this error.
+					continue
+				}
+				panic(err)
+				// TODO: handle EOF by redirection as specified at http://xrootd.org/doc/dev45/XRdv310.pdf, page 11
```

Comment:
is this comment/TODO still valid ?

Reply:
I have added a redirection here. However, not everything in a worst-case scenario is handled yet, so there is still a plenty of `TODO`s.